qiskit/transpiler/passes/synthesis/high_level_synthesis.py

```diff
@@ -160,84 +180,203 @@ def __init__(
-                methods = self.hls_config.methods[node.name]
-            elif (
-                self.hls_config.use_default_on_unspecified
-                and "default" in self.hls_plugin_manager.method_names(node.name)
-            ):
-                # the operation's name does not appear in the user-specified config,
-                # we use the "default" method when instructed to do so and the "default"
-                # method is available
-                methods = ["default"]
-            else:
-                methods = []
-
-            for method in methods:
-                # There are two ways to specify a synthesis method. The more explicit
-                # way is to specify it as a tuple consisting of a synthesis algorithm and a
-                # list of additional arguments, e.g.,
-                #   ("kms", {"all_mats": 1, "max_paths": 100, "orig_circuit": 0}), or
-                #   ("pmh", {}).
-                # When the list of additional arguments is empty, one can also specify
-                # just the synthesis algorithm, e.g.,
-                #   "pmh".
-                if isinstance(method, tuple):
-                    plugin_specifier, plugin_args = method
-                else:
-                    plugin_specifier = method
-                    plugin_args = {}
-
-                # There are two ways to specify a synthesis algorithm being run,
-                # either by name, e.g. "kms" (which then should be specified in entry_points),
-                # or directly as a class inherited from HighLevelSynthesisPlugin (which then
-                # does not need to be specified in entry_points).
-                if isinstance(plugin_specifier, str):
-                    if plugin_specifier not in self.hls_plugin_manager.method_names(node.name):
-                        raise TranspilerError(
-                            "Specified method: %s not found in available plugins for %s"
-                            % (plugin_specifier, node.name)
-                        )
-                    plugin_method = self.hls_plugin_manager.method(node.name, plugin_specifier)
-                else:
-                    plugin_method = plugin_specifier
 
-                qubits = (
-                    [dag.find_bit(x).index for x in node.qargs] if self._use_qubit_indices else None
-                )
+        # copy dag_op_nodes because we are modifying the DAG below
+        dag_op_nodes = dag.op_nodes()
 
-                decomposition = plugin_method.run(
-                    node.op,
-                    coupling_map=self._coupling_map,
-                    target=self._target,
-                    qubits=qubits,
-                    **plugin_args,
-                )
+        for node in dag_op_nodes:
+            qubits = (
+                [dag.find_bit(x).index for x in node.qargs] if self._use_qubit_indices else None
+            )
+            decomposition = self._recursively_handle_op(node.op, qubits)
```

Comment:
The question I have here is this means we're running the synthesis for every operation in the circuit right? Then if there is no synthesis `decomposition`  here is the same as `node.op` meaning we're essentially doing `dag.subsitute_node(node, node.op)` for every op node in the circuit? I'm wondering if it makes more sense to have `self._recursively_handle_op` return `None` in this case and we just do:

```suggestion
        for node in dag_op_nodes:
            qubits = (
                [dag.find_bit(x).index for x in node.qargs] if self._use_qubit_indices else None
            )
            decomposition = self._recursively_handle_op(node.op, qubits)
            if decomposition is None:
                continue
```

The other option I think would be to rebuild the dag from scratch (like we do in sabre and some other passes), as that's typically faster than doing in place substitution for every node. But in this case I expect the number of operations being synthesized are much less than those that aren't so just skipping the replacement should be sufficient.

Reply:
This is a very good point, and I was thinking of this in the follow-up PR #10965. The approach I took (also added here in [ba42605](https://github.com/Qiskit/qiskit/pull/9846/commits/ba42605f617e164e1f8beae19ce66b259bb465eb)) is to also return whether `op` was modified. It's slightly more convenient than returning `None`, for instance when we (recursively) synthesize the base gate of an annotated operation (which might itself be an annotated operation, hence the need for recursion) we don't need to check if the result is `None` and use `base.op` instead. 

Interesting that in cases rebuilding the DAG circuit from scratch is faster (but I agree that we don't need this here since in practice the number of high-level-objects / annotated gates will be smaller than the number of ordinary gates).